### PR TITLE
Open Clip multimodal embeddings

### PR DESCRIPTION
**Pull request by rlancemartin**

### CODE REVIEW

**Review by baskaryan — APPROVED** [review]

File: libs/langchain/langchain/embeddings/open_clip.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    def embed_documents(self, texts: List[str]) -> List[List[float]]:
+        text_features = [
+            self.model.encode_text(self.tokenizer(text)).tolist() for text in texts
+        ]
+        return text_features
+
+    def embed_query(self, text: str) -> List[float]:
+        return self.embed_documents([text])[0]
+
+    def embed_image(self, images: List[float]) -> List[List[float]]:
```
Comment (efriis) [inline libs/langchain/langchain/embeddings/open_clip.py:44]: The type on `images` seems like it should be a list of something 2d instead of floats

Comment (baskaryan) [reply libs/langchain/langchain/embeddings/open_clip.py:44]: good pt, in notebook seems like we're passing in list of np array even?

Comment (rlancemartin) [reply libs/langchain/langchain/embeddings/open_clip.py]: Yes, that's right. Image loading should go from compressed img file (`.JPEG`) -> decoded raw pixels w/ `PIL` -> `np.array`.

File: libs/langchain/langchain/embeddings/open_clip.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    def embed_image(self, images: List[np.ndarray]) -> List[List[float]]:
+        try:
+            from PIL import Image as _PILImage
+        except ImportError:
+            raise ImportError("Please install the PIL library: pip install pillow")
+        pil_images = [_PILImage.fromarray(image) for image in images]
+        image_features = [
+            self.model.encode_image(self.preprocess(pil_image).unsqueeze(0)).tolist()
+            for pil_image in pil_images
+        ]
```
Comment (efriis) [inline libs/langchain/langchain/embeddings/open_clip.py:55]: might want to optimize some of this conversion at some point, but works for now.

**Review by efriis — APPROVED** [review]

File: libs/langchain/langchain/embeddings/open_clip.py (hunk 2 above)
Comment (rlancemartin) [reply libs/langchain/langchain/embeddings/open_clip.py:55]: Ya, I'll need to look into that and also adding `_torch.no_grad():`. Will follow-up w/ those.

**Pull request merged**